Add tests for calculation of totalPrice

ChrisBrownie55 opened the pull request:
Built to solve "Total Price totals up all prices of all products properly" in #4
This PR extracts the logic of totalPrice to `util.js` allowing it to be tested separately from the rest of the hook logic.

### Changes made:

- refactored repeated setup for tests using `beforeEach`
- moved totaling and conversion logic to `util.js`
- handled when currency is undefined
- add tests for `calculateTotalPrice`:
 - adds all prices together
 - handles cents
 - handles different currencies
 - handles missing currencies


#### Other concerns

I do, however, have some concerns around the calculation of `totalPrice`. As I was looking at the code, I saw that you're inferring the *currency* based on the items inside the cart. This is fine unless it's possible that the items in the cart have different `currency` properties. That would cause the item in the cart to be improperly converted (1 to 1 instead of say 1 to 1.25 for GBP to USD). I don't have much experience with Stripe so this may just not be possible and therefore not an issue. If it is we can open up a new issue for that.

Let me know if I need to make any changes 👍

dayhaysoos commented:
@ChrisBrownie55 whoa! Thank you so much!

Honestly, the way things are added to the cart may need to be reflected on. The expectation is to never have different products with different currency, it's one of those things where I'm sure I'll never come across this issue and set it up this way rather than thinking of a way to avoid that possibliity altogether. 

This makes me think that there should be an issue to talk about the some default values and ways to let devs pass configs to this hook.

Perhaps in the context provider we can pass a "currency" value to refer to that instead of referring to what actually is getting passed?

ChrisBrownie55 commented:
So instead of hardcoding the default value to `'USD'` would the new version look akin to:

```jsx
  <CartProvider
    billingAddressCollection={...}
    successUrl={...}
    cancelUrl={...}
    stripe={...}
    currency="GBP"
  >
```

And then inside the `useStripeCart` hook:

```js
  const [cart, dispatch] = useContext(CartContext);


  const {
    /* other stuff */
    currency
  } = cart;

  const totalPrice = () => calculateTotalPrice(currency, cartItems);
```

And then for the logic of `calculateTotalPrice` should it be that only the passed in `currency` is used or should it still only use that as a fallback when `cartItem.currency` is `undefined`?

dayhaysoos commented:
To be safe, I think that we should make sure that it's only the passed in currency. The developer needs to determine what it is from the start. The intention is that these items will be coming from Stripe's API, so the currency needs to be determined before hand anyway

ChrisBrownie55 commented:
So would it be safe to assume that in all calls to `toCurrency`, the value passed as the `currency` should be replaced with the new one that's being passed down through `CartContext`?

dayhaysoos commented:
Yes, exactly. Thank you for the suggestion

ChrisBrownie55 commented:
# Update

I have added in the changes talked about in the aforementioned comments. To summarize, these changes are:

- Added `currency` prop to `CartProvider` component
- Now using that `currency` with `formatDetailedCart` and `calculateTotalPrice`
- `calculateTotalPrice` now uses only the provided `currency` prop from `CartProvider`
- Tests have been updated:
  - One of the tests for detailed SKUs was failing because it was adding an item with the currency of GBP and one with the currency of USD when the provided currency through `CartProvider` was USD. Therefore, the result is that the GBP was automatically converted to USD and thus failed the test.
  - Tests for `calculateTotalPrice` now reflect the change in how `currency` is used.

### Other changes

While I was modifying `formatDetailedCart`, I was looking at the reducer that was created and noticed a lot of redundancy and refactored it with the [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining) and [nullish coalescing](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Nullish_coalescing_operator) operators.

dayhaysoos commented:
@ChrisBrownie55 reviewing now 👁

dayhaysoos commented:
This is great! Thank you!
